shared/datastore.go

```diff
@@ -12,9 +16,40 @@ import (
 	"google.golang.org/appengine/datastore"
 )
 
+// Key abstracts an int64 based datastore.Key
+type Key interface {
+	IntID() int64
+}
+
+// Iterator abstracts a datastore.Iterator
+type Iterator interface {
+	Next(dst interface{}) (Key, error)
+}
+
+// Query abstracts a datastore.Query
+type Query interface {
+	Filter(filterStr string, value interface{}) Query
+	Project(project string) Query
+	Limit(limit int) Query
+	Offset(offset int) Query
+	Order(order string) Query
+	KeysOnly() Query
+	Distinct() Query
+	Run(Datastore) Iterator
```

Comment:
Shall we also add a `Run` function on `Datastore` for completeness? (The cloud Datastore package has its `Run` function on `Client` not `Query`.)

Reply:
I originally did, but, no; I'd rather be consistent with all of our datastore code so that copy-paste for similar queries will work, with the only difference being the way the datastore is created.